### DISCUSSION TITLE
How to set XmlFile Value "as is", meaning without XML encoding

### DISCUSSION THREAD

**Discussion by eykh** [wixtoolset/issues#7963] [Questions] [answered]
Hi,

is there a way to have `XmlFile` accept the contents of a MSI property as a value **without** encoding it to XML?
Our use case: We deploy our msi on a server and pass some XML content on the commandline. This content needs to be put into a certain XML element. This used to work on WIX V3 for years, when we used some config-file-search-n-replace extension for Wix. Well, this extension is no longer available for V4. 

The resulting XML file should look like this:
```
<AddIns>
    <add serverUrl="https://..." layerName="layer1" isDefault="true" displayName="bla"/>
    <add serverUrl="https://..." layerName="layer2" isDefault="false"/>
    <add serverUrl="https://..." layerName="layer3" isDefault="false" >
</AddIns>
```

But we get:
```
<AddIns>&lt;add serverUrl="https://..." layerName="layer1" isDefault="true" displayName="bla"/&gt;  ... </AddIns>
```

I guess you get the point. Do you have any advice, proposal, alternatives?

Any help is appreciated!

**Comment by robmen** [comment]
Sounds like an interesting feature to add.

**Comment by eykh** [comment]
Well, I see that the current implementation is **correct**, because it "protects" the XML from being messed up by weird string content. But if even you think, that this use case is something that could be added, I would rather propose to add something like a generic search-and-replace-in-template-file thingy. We used `SystemTools:TemplateFile` from [https://dblock.github.io/msiext/](https://dblock.github.io/msiext/), which seems to be no longer maintained and thus does not work with WIX V4.

This template file idea would be a catch-all solution (even for .ini files, whatever) and would not "pollute" the current `XmlFile` implementation.

Does this make sense?

**Reply by robmen** [reply]
Yes, that exists already. See @barnson's response.

**Comment by barnson** [comment, marked as answer]
https://wixtoolset.org/docs/schema/util/formatfile/

**Reply by eykh** [reply]
I was not aware of this! And I guess, most users aren't. I can't find any reference of this element on the net. The documentation also does not "ring any bell" for me, that this could be used to do search & replace or "templating" on arbitrary files types. 

Do you know of any link to an example usage of this element? Is the parent `File` element supposed to be a config file with formatted placeholders like `[SERVERPORT]` or so? What is the `BinaryRef` for then? Sorry, I don't get it, yet.

**Reply by barnson** [reply]
It's new to WiX v4. Yes, the file is formatted according to MSI formatting rules. The `BinaryRef` is a copy of the file, accessible at a time when data can be formatted.

**Reply by eykh** [reply]
Thank you! It looks like this is the perfect solution for template config files!
Could you please comment on how I used it?
```html
<Binary Id="binConfigTemplate" SourceFile="pathTo\myTemplate.config"/>
<Component Id="cmpConfigFile" Directory="INSTALLDIR">        
    <File Id="filConfigFile" Source="pathTo\myTemplate.config" KeyPath="yes">
        <util:FormatFile  BinaryRef="binConfigTemplate"/>                
    </File>
</Component>
``` 

The file `myTemplate.config` is of an arbitrary text format and contains [formatted](https://learn.microsoft.com/en-us/windows/win32/msi/formatted) content:

```
...
blabla
[MYMSIPROPERTY] bla bla
Better check the [SERVERURL]
...
``` 

It works, but do I use it the right way? I wonder if there's a way to avoid including the template file twice (as a `Binary` **and** as a `File`).
Any comments?
